File: Python/textbook_track/chapter05/ba5e.py
# Find a Highest-Scoring Alignment of Two Strings
import sys

from textbook_track.resources.utils import read_scoring_matrix


def collect_global_alignment(backtrack, string1, string2):
    ix = len(string1)
    jy = len(string2)
    aligned = []
    while ix > 0 or jy > 0:
        if ix == 0:
            aligned.append(('-', string2[jy-1]))
            jy -= 1
        elif jy == 0:
            aligned.append((string1[ix-1], '-'))
            ix -= 1
        elif backtrack[ix-1][jy-1] == 0:
            aligned.append((string1[ix-1], string2[jy-1]))
            ix -= 1
            jy -= 1
        elif backtrack[ix-1][jy-1] == -1:
            aligned.append((string1[ix-1], '-'))
            ix -= 1
        else:
            aligned.append(('-', string2[jy-1]))
            jy -= 1

    return map(''.join, zip(*aligned[::-1]))


def calc_global_alignment(string1, string2, scoring_matrix, sigma):
    n = len(string1)
    m = len(string2)
    longest_path = [[0] * (m+1) for _ in range(n+1)]
    backtrack = [[0] * m for _ in range(n)]
    for ix in range(1, n+1):
        longest_path[ix][0] = -ix * sigma
    for jy in range(1, m+1):
        longest_path[0][jy] = -jy * sigma

    for ix, c1 in enumerate(string1):
        for jy, c2 in enumerate(string2):
            deletion = longest_path[ix][jy+1] - sigma
            insertion = longest_path[ix+1][jy] - sigma
            match = longest_path[ix][jy] + scoring_matrix[(c1, c2)]
            longest_path[ix+1][jy+1] = max(insertion, deletion, match)

            if longest_path[ix+1][jy+1] == deletion:
                backtrack[ix][jy] = -1
            elif longest_path[ix+1][jy+1] == insertion:
                backtrack[ix][jy] = 1

    aligned_string1, aligned_string2 = collect_global_alignment(backtrack, string1, string2)
    return longest_path[n][m], aligned_string1, aligned_string2


def main():
    data = sys.stdin.read().splitlines()
    string1 = data[0]
    string2 = data[1]
    blosum62 = read_scoring_matrix()
    sigma = 5
    score, aligned_string1, aligned_string2 = calc_global_alignment(string1, string2, blosum62, sigma)
    print(score)
    print(aligned_string1)
    print(aligned_string2)


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter05/BA5E.scala
package TextbookTrack.Chapter05

object BA5E {
  import scala.annotation.tailrec
  import TextbookTrack.Utils.readScoringMatrix

  private val Blosum62: Map[(Char, Char), Int] = readScoringMatrix("BLOSUM62")

  private def collectGlobalAlignment(backtrack: Array[Array[Byte]],
                                     string1: String,
                                     string2: String): (String, String) = {
    @tailrec
    def loop(aligned: List[(Char, Char)], ix: Int, jy: Int): (String, String) = {
      if (ix == 0 && jy == 0) {
        val (aligned1, aligned2): (List[Char], List[Char]) = aligned.unzip
        (aligned1.mkString, aligned2.mkString)
      }
      else if (ix == 0) loop(('-', string2(jy - 1)) :: aligned, ix, jy - 1)
      else if (jy == 0) loop((string1(ix - 1), '-') :: aligned, ix - 1, jy)
      else if (backtrack(ix - 1)(jy - 1) == 0) loop((string1(ix - 1), string2(jy - 1)) :: aligned, ix - 1, jy - 1)
      else if (backtrack(ix - 1)(jy - 1) == -1) loop((string1(ix - 1), '-') :: aligned, ix - 1, jy)
      else loop(('-', string2(jy - 1)) :: aligned, ix, jy - 1)
    }

    loop(Nil, string1.length, string2.length)
  }

  def calcGlobalAlignment(string1: String, string2: String, sigma: Int): (Int, String, String) = {
    val n: Int = string1.length
    val m: Int = string2.length
    val table: Array[Array[Int]] = Array.fill(n + 1, m + 1)(0)
    val backtrack: Array[Array[Byte]] = Array.fill(n, m)(0)
    (1 to n).foreach(ix => table(ix)(0) = -ix * sigma)
    (1 to m).foreach(jy => table(0)(jy) = -jy * sigma)

    for {
      (c1, ix) <- string1.zipWithIndex
      (c2, jy) <- string2.zipWithIndex
    } {
      val deletion: Int = table(ix)(jy + 1) - sigma
      val insertion: Int = table(ix + 1)(jy) - sigma
      val matching: Int = table(ix)(jy) + Blosum62((c1, c2))
      val maxPath: Int = List(deletion, insertion, matching).max
      table(ix + 1)(jy + 1) = maxPath
      backtrack(ix)(jy) = if (maxPath == deletion) -1 else if (maxPath == insertion) 1 else 0
    }

    val (alignedString1, alignedString2): (String, String) = collectGlobalAlignment(backtrack, string1, string2)
    (table(n)(m), alignedString1, alignedString2)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val string1: String = reader.next()
    val string2: String = reader.next()
    val sigma: Int = 5
    val (score, alignedString1, alignedString2): (Int, String, String) = calcGlobalAlignment(string1, string2, sigma)
    println(score)
    println(alignedString1)
    println(alignedString2)
  }
}
